src/ai/defaultai_seafaring.cc

```diff
@@ -212,27 +206,38 @@ bool DefaultAI::marine_main_decisions(const Time& gametime) {
 		}
 	}
 
-	// building a ship? if yes, find a shipyard and order it to build a ship
-	if (enough_ships == FleetStatus::kNeedShip) {
-
-		for (const ProductionSiteObserver& ps_obs : productionsites) {
-			if (ps_obs.bo->is(BuildingAttribute::kShipyard) && ps_obs.site->can_start_working() &&
-			    ps_obs.site->is_stopped()) {
-				// make sure it is fully stocked
-				// counting stocks
-				uint8_t stocked_wares = 0;
-				std::vector<Widelands::InputQueue*> const inputqueues = ps_obs.site->inputqueues();
-				for (Widelands::InputQueue* queue : inputqueues) {
-					if (queue->get_type() == Widelands::wwWARE) {
-						stocked_wares += queue->get_filled();
-					}
-				}
-				if (stocked_wares < 16) {
-					continue;
+	// goes over productionsites finds shipyards and configures them
+	for (const ProductionSiteObserver& ps_obs : productionsites) {
+		if (ps_obs.bo->is(BuildingAttribute::kShipyard)) {
+			// counting stocks
+			uint8_t stocked_wares = 0;
+			std::vector<Widelands::InputQueue*> const inputqueues = ps_obs.site->inputqueues();
+			for (Widelands::InputQueue* queue : inputqueues) {
+				if (queue->get_type() == Widelands::wwWARE) {
+					stocked_wares += queue->get_filled();
 				}
+			}
+			if (stocked_wares == 16 && ps_obs.site->is_stopped() && ps_obs.site->can_start_working()) {
+				idle_shipyard_stocked = true;
+			}
 
+			if (expeditions_ready > 0 && idle_shipyard_stocked && allships.size() - ports_count == 0) {
```

Comment:
this is describing the condition when a shipyard potentially is being built in the wrong place. unfortunately we need to wait for at least one ship build there to detect this.
background: if a port exists and a shipyard exists a ship is build as long as the # of available ships is lower then the number of ports. If an expedition is ready a ship is assigned and will move to the port and take the expedition resulting in no expedition rerady at the Protdock (pd). 
so if we have a port, a ship and an expedition ready we either have the ship in the wrong sea or the ship is underway so we repeat the check in 2 consecutive decision intervals and if it persists (determined from the variable "potential_wrong_shipyard" we dismantle the shipyard and sink the ship.